cache/in-memory/src/permission.rs

```diff
@@ -0,0 +1,434 @@
+    ///
+    /// // later on...
+    ///
+    /// let guild_id = GuildId(4);
+    /// let user_id = UserId(5);
+    ///
+    /// let permissions = cache.permissions().root(user_id, guild_id)?;
+    /// println!(
+    ///     "User {} in guild {} has permissions {:?}",
+    ///     user_id,
+    ///     guild_id,
+    ///     permissions,
+    /// );
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`RootErrorType::MemberUnavailable`] error type if the
+    /// member for the user in the guild is not present.
+    ///
+    /// Returns a [`RootErrorType::RoleUnavailable`] error type if one of the
+    /// member's roles is not in the cache.
+    ///
+    /// [`ResourceType::MEMBER`]: crate::ResourceType::MEMBER
+    /// [`ResourceType::ROLE`]: crate::ResourceType::ROLE
+    /// [`ResourceType`]: crate::ResourceType
+    pub fn root(&self, user_id: UserId, guild_id: GuildId) -> Result<Permissions, RootError> {
+        let MemberRoles { assigned, everyone } = self
+            .member_roles(user_id, guild_id)
+            .map_err(RootError::from_member_roles)?;
+        let calculator =
+            PermissionCalculator::new(guild_id, user_id, everyone, assigned.as_slice());
+
+        Ok(calculator.root())
+    }
+
+    /// Retrieve a member's roles' permissions and the guild's `@everyone`
+    /// role's permissions.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`MemberRolesErrorType::MemberMissing`] if the member is missing
+    /// from the cache.
+    ///
+    /// Returns [`MemberRolesErrorType::RoleMissing`] if a role is missing from
+    /// the cache.
+    fn member_roles(
+        &self,
+        user_id: UserId,
+        guild_id: GuildId,
+    ) -> Result<MemberRoles, MemberRolesErrorType> {
+        let member = if let Some(member) = (self.0).0.members.get(&(guild_id, user_id)) {
+            member
+        } else {
+            return Err(MemberRolesErrorType::MemberMissing { guild_id, user_id });
+        };
+
+        let mut member_roles = Vec::with_capacity(member.roles.len());
+
+        for role_id in &member.roles {
+            let role = if let Some(role) = (self.0).0.roles.get(role_id) {
+                role
+            } else {
+                return Err(MemberRolesErrorType::RoleMissing { role_id: *role_id });
+            };
+
+            member_roles.push((*role_id, role.data.permissions));
+        }
+
+        // Assume that the `@everyone` role is always present, so do this last.
+        let everyone_role_id = RoleId(guild_id.0);
+
+        if let Some(everyone_role) = (self.0).0.roles.get(&everyone_role_id) {
+            Ok(MemberRoles {
+                assigned: member_roles,
+                everyone: everyone_role.data.permissions,
+            })
+        } else {
+            Err(MemberRolesErrorType::RoleMissing {
+                role_id: everyone_role_id,
+            })
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Do we want to test these methods, or are they already covered by tests in the other PR?

Reply:
Addressed